File: apps/frontend/src/adapters/shared/dividends.ts
import { invoke, logger } from "./platform";

export type DividendCalendarEventKind =
  | "posted"
  | "past_unposted"
  | "upcoming_estimated";

export interface DividendCalendarEvent {
  id: string;
  date: string;
  symbol: string;
  accountId: string;
  accountName: string;
  displayAmount: string | number;
  currency: string;
  kind: DividendCalendarEventKind;
  activityId?: string | null;
  notes?: string | null;
}

export interface AccountDividendSettings {
  enabled: boolean;
  dividendTaxRate: number;
}

export interface DividendSyncSettings {
  globalEnabled: boolean;
  accounts: Record<string, AccountDividendSettings>;
}

export interface DividendSyncAccountResult {
  accountId: string;
  accountName: string;
  created: number;
  skippedNoShares: number;
  skippedDuplicates: number;
}

export interface DividendSyncResult {
  created: number;
  skipped: number;
  skippedNoShares: number;
  skippedDuplicates: number;
  accounts: DividendSyncAccountResult[];
  errors: string[];
  netCashAdded: string;
}

export interface AssetDividendView {
  assetId: string;
  symbol: string;
  currency: string;
  ytdIncome: string | number;
  ttmIncome: string | number;
  events: DividendCalendarEvent[];
}

export const getDividendSyncSettings = async (): Promise<DividendSyncSettings> => {
  try {
    return await invoke<DividendSyncSettings>("get_dividend_sync_settings");
  } catch (error) {
    logger.error("Error fetching dividend sync settings.");
    throw error;
  }
};

export const updateDividendSyncSettings = async (
  settings: DividendSyncSettings,
): Promise<DividendSyncSettings> => {
  try {
    return await invoke<DividendSyncSettings>("update_dividend_sync_settings", { settings });
  } catch (error) {
    logger.error("Error updating dividend sync settings.");
    throw error;
  }
};

export const syncDividends = async (): Promise<DividendSyncResult> => {
  try {
    return await invoke<DividendSyncResult>("sync_dividends");
  } catch (error) {
    logger.error("Error syncing dividends.");
    throw error;
  }
};

export const removeAutoDividends = async (): Promise<number> => {
  try {
    return await invoke<number>("remove_auto_dividends");
  } catch (error) {
    logger.error("Error removing auto dividends.");
    throw error;
  }
};

export const getDividendCalendarEvents = async (): Promise<DividendCalendarEvent[]> => {
  try {
    return await invoke<DividendCalendarEvent[]>("get_dividend_calendar_events");
  } catch (error) {
    logger.error("Error fetching dividend calendar events.");
    throw error;
  }
};

export const getAssetDividendView = async (assetId: string): Promise<AssetDividendView> => {
  try {
    return await invoke<AssetDividendView>("get_asset_dividend_view", { assetId });
  } catch (error) {
    logger.error("Error fetching asset dividend view.");
    throw error;
  }
};

File: apps/tauri/src/commands/dividends.rs
use std::sync::Arc;

use crate::context::ServiceContext;
use log::debug;
use tauri::State;
use wealthfolio_core::dividends::{
    AssetDividendView, DividendCalendarEvent, DividendSyncResult, DividendSyncSettings,
};

#[tauri::command]
pub async fn get_dividend_sync_settings(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<DividendSyncSettings, String> {
    debug!("Fetching dividend sync settings...");
    state
        .dividend_sync_service()
        .get_settings()
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn update_dividend_sync_settings(
    settings: DividendSyncSettings,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<DividendSyncSettings, String> {
    debug!("Updating dividend sync settings...");
    state
        .dividend_sync_service()
        .update_settings(settings)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn sync_dividends(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<DividendSyncResult, String> {
    debug!("Syncing market-data dividends...");
    state
        .dividend_sync_service()
        .sync()
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn sync_dividends_account(
    account_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<DividendSyncResult, String> {
    debug!("Syncing market-data dividends for {}...", account_id);
    state
        .dividend_sync_service()
        .sync_account(&account_id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn remove_auto_dividends(state: State<'_, Arc<ServiceContext>>) -> Result<usize, String> {
    debug!("Removing auto-created dividends...");
    state
        .dividend_sync_service()
        .remove_auto_created()
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn remove_auto_dividends_account(
    account_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<usize, String> {
    debug!("Removing auto-created dividends for {}...", account_id);
    state
        .dividend_sync_service()
        .remove_auto_created_account(&account_id)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_dividend_calendar_events(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<DividendCalendarEvent>, String> {
    debug!("Building dividend calendar events...");
    state
        .dividend_sync_service()
        .build_calendar_events()
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_asset_dividend_view(
    asset_id: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<AssetDividendView, String> {
    debug!("Building asset dividend view for {}...", asset_id);
    state
        .dividend_sync_service()
        .build_asset_dividend_view(&asset_id)
        .await
        .map_err(|e| e.to_string())
}
